File: samples/Android/FXTSMobileTableManager/app/src/main/java/com/FXTSMobile/CommissionsCalculator.java
package com.FXTSMobile;

import com.fxcore2.*;

import java.util.Map;

public class CommissionsCalculator {

    private final String COMMISSIONED_UI_ENABLED_PROP = "COMMISSIONED_UI_ENABLED";

    private O2GCommissionsProvider mCommissionsProvider;
    private O2GTradingSettingsProvider mSettingsProvider;

    private O2GAccountsTable mAccountsTable;
    private O2GAccountTableRow mAccountTableRow;

    private boolean mIsCommEnabled;

    public CommissionsCalculator(O2GSession session) {
        this(session, null);

        if (mAccountsTable.size() > 0) {
            mAccountTableRow = mAccountsTable.getRow(0);
        }
    }

    public CommissionsCalculator(O2GSession session, O2GAccountTableRow accountTableRow) {
        if (session.getSessionStatus() != O2GSessionStatusCode.CONNECTED)
            throw new IllegalStateException("session_not_ready");

        mAccountTableRow = accountTableRow;

        initializeTables();
        initializeCommProvider(session);

        mIsCommEnabled = detectCommissionsSupport(session);
    }

    private void initializeTables() {
        mAccountsTable = (O2GAccountsTable) SharedObjects.getInstance()
                .getSession().getTableManager().getTable(O2GTableType.ACCOUNTS);
    }

    private void initializeCommProvider(O2GSession session) {
        mCommissionsProvider = session.getCommissionsProvider();

        while (mCommissionsProvider.getStatus() != O2GCommissionStatusCode.READY &&
                mCommissionsProvider.getStatus() != O2GCommissionStatusCode.FAIL &&
                mCommissionsProvider.getStatus() != O2GCommissionStatusCode.DISABLED) {
            try {
                Thread.sleep(50);
            } catch (InterruptedException e) {
                throw new IllegalStateException("commissions_provider_not_ready", e);
            }
        }

        if (mCommissionsProvider.getStatus() != O2GCommissionStatusCode.READY)
            throw new IllegalStateException("commissions_provider_not_ready");
    }

    private boolean detectCommissionsSupport(O2GSession session) {
        O2GLoginRules loginRules = session.getLoginRules();
        mSettingsProvider = loginRules.getTradingSettingsProvider();
        O2GResponse response = loginRules.getSystemPropertiesResponse();

        O2GSystemPropertiesReader reader = session.getResponseReaderFactory().createSystemPropertiesReader(response);
        if (reader == null)
            return false;

        Map<String, String> properties = reader.getProperties();
        if (properties == null)
            return false;

        if (properties.containsKey(COMMISSIONED_UI_ENABLED_PROP)) {
            String commPropValue = properties.get(COMMISSIONED_UI_ENABLED_PROP);
            return commPropValue != null && commPropValue.equals("Y");
        }

        return false;
    }

    private int calcAmount(O2GOfferTableRow offersTableRow, int lots) {
        int baseUnitSize = mSettingsProvider.getBaseUnitSize(offersTableRow.getInstrument(), mAccountTableRow);
        return baseUnitSize * lots;
    }

    // public
    //

    public double calcOpenCommission(O2GOfferTableRow offersTableRow, int lots, boolean isBuy) {
        if (mAccountTableRow == null || offersTableRow == null)
            return Double.NaN;

        int amount = calcAmount(offersTableRow, lots);
        String buySell = isBuy ? "B" : "S";
        return mCommissionsProvider.calcOpenCommission(offersTableRow, mAccountTableRow, amount, buySell, 0);
    }

    public double calcCloseCommission(O2GOfferTableRow offersTableRow, int lots, boolean isBuy) {
        if (mAccountTableRow == null || offersTableRow == null)
            return Double.NaN;

        int amount = calcAmount(offersTableRow, lots);
        String buySell = isBuy ? "B" : "S";
        return mCommissionsProvider.calcCloseCommission(offersTableRow, mAccountTableRow, amount, buySell, 0);
    }

    public double calcTotalCommission(O2GOfferTableRow offersTableRow, int lots, boolean isBuy) {
        if (mAccountTableRow == null || offersTableRow == null)
            return Double.NaN;

        int amount = calcAmount(offersTableRow, lots);
        String buySell = isBuy ? "B" : "S";
        return mCommissionsProvider.calcTotalCommission(offersTableRow, mAccountTableRow, amount, buySell, 0, 0);
    }

    public boolean isCommEnabled() {
        return mIsCommEnabled;
    }
}

File: samples/iOS/iTradeStation(TM)/iTradeStation(TM)/CommissionsCalculator.swift
import Foundation

class CommissionsCalculator {
    private let COMMISSIONED_UI_ENABLED_PROP = "COMMISSIONED_UI_ENABLED"

    private var session: IO2GSession
    private var settingsProvider: IO2GTradingSettingsProvider?
    private var offersTable: IO2GOffersTable?
    private var accountTableRow: IO2GAccountTableRow?
    
    // the commissionsProvider will be initialized in the OffersViewController instance
    private lazy var commissionsProvider: IO2GCommissionsProvider = initializeCommProvider()
    
    private var _isCommEnabled: Bool = false
    
    public var isCommEnabled : Bool {
        get {
            return _isCommEnabled
        }
    }
    
    init(session: IO2GSession, accountTableRow: IO2GAccountTableRow?, offersTable: IO2GOffersTable) {
        if session.getStatus() != IO2GSession_Connected {
            fatalError("session_not_ready")
        }
        
        self.session = session
        self.offersTable = offersTable
        self.accountTableRow = accountTableRow;
        self._isCommEnabled = detectCommissionsSupport(session: session)
    }
    
    private func detectCommissionsSupport(session: IO2GSession) -> Bool {
        var loginRules = session.getLoginRules()
        while loginRules == nil { loginRules = session.getLoginRules() }
        self.settingsProvider = loginRules!.getTradingSettingsProvider()
        let response = loginRules!.getSystemPropertiesResponse()
        if response == nil {
            return false
        }
        let reader = session.getResponseReaderFactory().createSystemPropertiesReader(response);
        if reader == nil {
            return false
        }
        let commEnabledVal = reader?.findProperty(COMMISSIONED_UI_ENABLED_PROP)
        if commEnabledVal == nil {
            return false
        }
        
        return commEnabledVal! == "Y";
    }
    
    private func initializeCommProvider() -> IO2GCommissionsProvider {
        let commissionsProvider = session.getCommissionsProvider()
        var downCounter = 25
        while (commissionsProvider!.getStatus() != CommissionStatusReady && downCounter > 0)
        {
            if commissionsProvider!.getStatus() == CommissionStatusDisabled { break }
            if commissionsProvider!.getStatus() == CommissionStatusFailToLoad { break }
            usleep(50000) // 50 millis
            downCounter -= 1
        }
        return commissionsProvider!
    }
    
    private func findOffersRow(offerID: String) -> IO2GOfferTableRow? {
        return offersTable!.findRow(offerID)
    }
    
    private func calcAmount(offersTableRow: IO2GOfferTableRow, lots: Int32) -> Int32 {
        let baseUnitSize = settingsProvider!.getBaseUnitSize(offersTableRow.getInstrument(), accountTableRow);
        return baseUnitSize * lots;
    }
    
    public func calcOpenCommission(offerID: String, lots: Int32, isBuy: Bool) -> Double {
        
        let offersTableRow = findOffersRow(offerID: offerID)
        
        if accountTableRow == nil || offersTableRow == nil {
            return Double.nan
        }
        
        let amount = calcAmount(offersTableRow: offersTableRow!, lots: lots)
        let buySell = isBuy ? "B" : "S"
        return commissionsProvider.calcOpenCommission(offersTableRow, accountTableRow, amount, buySell, 0);
    }
    
    public func calcCloseCommission(offerID: String, lots: Int32, isBuy: Bool) -> Double {
        
        let offersTableRow = findOffersRow(offerID: offerID)
        
        if accountTableRow == nil || offersTableRow == nil {
            return Double.nan
        }
        
        let amount = calcAmount(offersTableRow: offersTableRow!, lots: lots)
        let buySell = isBuy ? "B" : "S"
        return commissionsProvider.calcCloseCommission(offersTableRow, accountTableRow, amount, buySell, 0);
    }
    
    public func calcTotalCommission(offerID: String, lots: Int32, isBuy: Bool) -> Double {
        
        let offersTableRow = findOffersRow(offerID: offerID)
        
        if accountTableRow == nil || offersTableRow == nil {
            return Double.nan
        }
        
        let amount = calcAmount(offersTableRow: offersTableRow!, lots: lots)
        let buySell = isBuy ? "B" : "S"
        return commissionsProvider.calcTotalCommission(offersTableRow, accountTableRow, amount, buySell, 0, 0);
    }
}
